src/hb-cairo.cc

```diff
@@ -0,0 +1,867 @@
+ * factor defines how the scale set on such #hb_font_t objects
+ * relates to the font-matrix (as such font size) of the cairo
+ * scaled-font.
+ *
+ * If the scale-factor is zero (default), then the scale of the
+ * #hb_font_t object will be left at default, which is the UPEM
+ * value of the respective #hb_face_t.
+ *
+ * If the scale-factor is set to non-zero, then the X and Y scale
+ * of the #hb_font_t object will be respectively set to the
+ * @scale_factor times the xx and yy elements of the scale-matrix
+ * of the cairo scaled-font being created.
+ *
+ * When using the hb_cairo_glyphs_from_buffer() API to convert
+ * HarfBuzz glyph buffer resulted from shaping with such a hb_font_t,
+ * if the scale-factor was non-zero, you can pass it directly to
+ * that API.  If the scale-factor was zero however, you need to
+ * calculate the correct scale-factor to pass to that API by
+ * dividing the #hb_font_t scale-factor by the cairo scaled-font's
+ * scale-matrix scale-factor. This assumes a uniform scale.
+ *
+ * If the cairo-face was created using the alternative constructor
+ * hb_cairo_font_face_create_for_font(), you are on your own
+ * computing the correct scale-factor to pass to
+ * hb_cairo_glyphs_from_buffer(), but it is generally the x_scale
+ * of the #hb_font_t divided by the xx factor of the scaled-font's
+ * scale-matrix.
+ *
+ * Since: REPLACEME
+ */
+void
+hb_cairo_font_face_set_scale_factor (cairo_font_face_t *font_face,
+				     unsigned int scale_factor)
+{
+  cairo_font_face_set_user_data (font_face,
+				 &hb_cairo_scale_factor_user_data_key,
+				 (void *) (uintptr_t) scale_factor,
+				 nullptr);
+}
+
+/**
+ * hb_cairo_font_face_get_scale_factor:
+ * @font_face: a `cairo_font_face_t`
+ *
+ * Gets the scale factor set on the @font_face. Defaults to zero.
+ * See hb_cairo_font_face_set_scale_factor() for details.
+ *
+ * Returns: the scale factor of @font_face
+ *
+ * Since: REPLACEME
+ */
+unsigned int
+hb_cairo_font_face_get_scale_factor (cairo_font_face_t *font_face)
+{
+  return (unsigned int) (uintptr_t)
+	 cairo_font_face_get_user_data (font_face,
+					&hb_cairo_scale_factor_user_data_key);
+}
+
+
+/**
+ * hb_cairo_glyphs_from_buffer:
+ * @buffer: a `hb_buffer_t` containing glyphs
+ * @utf8_clusters: `true` to if @buffer clusters are in bytes, instead of characters
+ * @scale_factor: scale factor to divide hb_positions_t values by
+ * @x: X position to place first glyph
+ * @y: Y position to place first glyph
+ * @utf8: (nullable): the text that was shaped in @buffer
+ * @utf8_len: the length of @utf8 in bytes
+ * @glyphs: (out): return location for an array of `cairo_glyph_t`
+ * @num_glyphs: (out): return location for the length of @glyphs
+ * @clusters: (out) (nullable): return location for an array of cluster positions
+ * @num_clusters: (out) (nullable): return location for the length of @clusters
+ * @cluster_flags: (out) (nullable): return location for cluster flags
```

Comment:
The out arrays are allocated using cairo function, do they need to be deallocated with similar functions as well, if so, then this should be documented.